sdk/rust/src/contract/contract_function_parameters.rs

```diff
@@ -0,0 +1,1268 @@
+            .append(&mut Self::left_pad_32_bytes(elements_len.to_be_bytes().as_slice(), false));
+        let mut current_offset = elements_len * 32;
+        let mut body_bytes: Vec<u8> = Vec::new();
+        for element in elements {
+            let element = element.as_ref();
+            out_bytes.append(&mut Self::left_pad_32_bytes(
+                current_offset.to_be_bytes().as_slice(),
+                false,
+            ));
+            current_offset += element.len();
+            body_bytes.extend(element)
+        }
+        out_bytes.extend(body_bytes);
+        out_bytes
+    }
+
+    fn encode_array_of_32_byte_elements<I>(elements: I, elements_len: usize) -> Vec<u8>
+    where
+        I: IntoIterator<Item = Vec<u8>>,
+    {
+        let mut out_bytes = Self::left_pad_32_bytes(elements_len.to_be_bytes().as_slice(), false);
+        out_bytes.reserve(out_bytes.len() + (elements_len * 32));
+        out_bytes.extend(elements.into_iter().flatten());
+        out_bytes
+    }
+
+    pub(crate) fn to_bytes(&self, func_name: Option<&str>) -> Vec<u8> {
+        let mut current_dynamic_offset = self.args.len() * 32;
+        let mut arg_bytes = Vec::new();
+        let mut dynamic_arg_bytes = Vec::new();
+        let mut function_selector = func_name.map(ContractFunctionSelector::new);
+        for arg in &self.args {
+            if let Some(selector) = &mut function_selector {
+                selector.add_param_type(arg.type_name);
+            }
+            if arg.is_dynamic {
+                arg_bytes.append(&mut Self::left_pad_32_bytes(
+                    current_dynamic_offset.to_be_bytes().as_slice(),
+                    false,
+                ));
+                dynamic_arg_bytes.extend_from_slice(arg.value_bytes.as_slice());
+                current_dynamic_offset += arg.value_bytes.len();
+            } else {
+                arg_bytes.extend_from_slice(arg.value_bytes.as_slice());
+            }
+        }
+        arg_bytes.append(&mut dynamic_arg_bytes);
+        if let Some(selector) = &mut function_selector {
+            let mut out_bytes = Vec::from(selector.finish());
+            out_bytes.append(&mut arg_bytes);
+            out_bytes
+        } else {
+            arg_bytes
+        }
+    }
+
+    /// Add a `string` argument to the `ContractFunctionParameters`
+    #[allow(dead_code)]
+    pub fn add_string(&mut self, val: &str) -> &mut Self {
+        self.args.push(Argument {
+            type_name: "string",
+            value_bytes: Self::encode_dynamic_bytes(val.as_bytes()),
+            is_dynamic: true,
+        });
+        self
+    }
+
+    /// Add a `string[]` argument to the `ContractFunctionParameters`
+    #[allow(dead_code)]
+    pub fn add_string_array(&mut self, val: &[&str]) -> &mut Self {
+        self.args.push(Argument {
+            type_name: "string[]",
+            value_bytes: Self::encode_array_of_dynamic_byte_arrays(
+                val.into_iter().map(|s| s.as_bytes()),
```

Comment:
`&str` actually implements `AsRef<[u8]>`, but that's kinda moot because this function should take `&[T]` where `T: AsRef<str>` because you want to be able to accept things like `&[String]`, `&[Box<str>]`, `&[TinyStr]` (which is a type that we "don't know about" but carry about anyway), `&[Cow<'static, str>]` etc. So, we'd still need to `map` here (`.map(T::as_ref)`).